src/main.c

```diff
@@ -603,6 +604,83 @@ static ros_reply_t *receive_reply (ros_connection_t *c) /* {{{ */
 	return (head);
 } /* }}} ros_reply_t *receive_reply */
 
+static int connect_socket_timeout (struct addrinfo *ai_ptr, unsigned int timeout_sec)
+{
+	struct timeval timeout = {
+		.tv_sec = timeout_sec,
+	};
+
+	/* value of 0 means no timeout, select treats NULL as no timeout */
+	struct timeval *timeout_ptr = (0 == timeout_sec)? NULL : &timeout;
+
+	int fd;
+	int status;
+
+	fd = socket (ai_ptr->ai_family, ai_ptr->ai_socktype, ai_ptr->ai_protocol);
+	if (fd < 0)
+	{
+		ros_debug ("connect_socket_timeout: socket(2) failed.\n");
+		return (0);
+	}
+
+	/* set socket nonblocking just for the connect() - allows us to call select() on connecting socket */
+	fcntl(fd, F_SETFL, O_NONBLOCK);
+
+	status = connect (fd, ai_ptr->ai_addr, ai_ptr->ai_addrlen);
+	if (status != 0 && (errno != EINPROGRESS))
+	{
+		/* connect() failed due to some reason other than a timeout */
+		ros_debug ("connect_socket_timeout: connect(2) failed.\n");
+		close (fd);
+		return (0);
+	}
+
+	fd_set fdset;
+	FD_ZERO (&fdset);
+	FD_SET (fd, &fdset);
+
+	if (1 == select (fd + 1, NULL, &fdset, NULL, timeout_ptr))
+	{
+		/* find out what happened to the socket */
+		int socket_error;
+		socklen_t len = sizeof(socket_error);
+
+		if (0 != getsockopt (fd, SOL_SOCKET, SO_ERROR, &socket_error, &len))
+		{
+			ros_debug ("connect_socket_timeout: getsockopt() failed\n");
+			close (fd);
+			return (0);
+		}
+
+		if (0 != socket_error)
+		{
+			ros_debug ("connect_socket_timeout: connect() failed.\n");
+			close (fd);
+
+			/* fill in errno so the caller knows what happened */
+			errno = socket_error;
```

Comment:
I'm not too confident about the practice of assigning to errno, but I can't think of any other way to do it without changing the API of the library. Suggestions welcome!

Reply:
I think there is nothing bad in assigning to errno.
There are many uses of such in libraries, including same as `errno = socket_error` assignment.

Example implementation: https://github.com/redis/hiredis/blob/master/net.c#L235 (two functions starting from L235 and below).
I liked their API of error reporting, it may be used as reference when extending librouteros.

Reply:
Thanks for the feedback + glad to know this practice is being used elsewhere.